File: matlab/segop.m
function SOC = segop(k,Q,x0)

% function SOC = segop(k,Q,xO). This function returns the
% parameters for the segmentally optimal composite curve.
% It receives the .IG curve parameters, x0, data points Q,
% and knot sequence k. It was written by E. J. Lane.

[P,ang,dt] = ktangdt(x0);   % Separates the vector xO
                            % into its subcomponents.
                            
disp("P");
disp(P);
disp("ang");
disp(ang);
disp("dt");
disp(dt);          
disp("Q");
disp(Q);               
bdt = bstdst(dt,Q,P,ang,k); % Call to the function which finds
                            % the optimum distances for a segment.
disp("bdt");
disp(bdt);                      
                            
bdt1 = [];                  % ECR

for i = 1 : 2 % Loop to assemble the "best" distances.
    bdt1 =  [bdt1 bdt(i,:)];
end

% Assemble the vector of parameters for the curve.
SOC = [P(1,:) P(2,:) ang bdt1];

File: scripts/segop.py
import numpy as np
from ktangdt import ktangdt
from bstdst import bstdst

def segop(k, Q, x0):
    # Separates the vector x0 into its subcomponents.
    # print("k: ", k)
    # print("x0: ", x0)

    P, ang, dt = ktangdt(x0)

    print("P: ", P)
    print("ang: ", ang)
    print("dt: ", dt)
    print("Q: ", Q)
    # Call to the function which finds the optimum distances for a segment.
    bdt = bstdst(dt, Q, P, ang, k)

    print("bdt: ", bdt)
    bdt1 = [] # Empty list
    # PROBLEMA TA ACIMA DAQUI bdt VEM ERRADO

    for i in range(2):
        bdt1.extend(bdt[i, :])

    # Assemble the vector of parameters for the curve.
    SOC = np.concatenate([P[0, :], P[1, :], ang, bdt1])
    return SOC